### ISSUE TITLE
IOS build fails when using SSL enable libwebsockets

### ISSUE THREAD

**Issue by div5yesh** [cocos2d/cocos2d-x#17723]
- cocos2d-x version: 3.14
- devices test on: iphone simulator
- developing environments
   - NDK version: NA
   - Xcode version: 8.3.1
   - VS version: NA
   - browser type and version: NA

Steps to Reproduce:

1. cocos new testproject -l js
2. ./build.sh -p=ios --libs=websockets --arch=i386 --mode=debug
3. copy built websocket libraries to testproject
4. cocos compile -p ios
5. open testproject.xcodeproj
6. build and run project

**Undefined symbols for architecture x86_64**:
 ```
 "_lws_remaining_packet_payload", referenced from:
      cocos2d::network::WebSocket::onClientReceivedData(void*, long) in libcocos2d iOS.a(WebSocket.o)
  "_lws_write", referenced from:
      cocos2d::network::WebSocket::onClientWritable() in libcocos2d iOS.a(WebSocket.o)
  "_lws_get_context", referenced from:
      cocos2d::network::WebSocketCallbackWrapper::onSocketCallback(lws*, lws_callback_reasons, void*, void*, unsigned long) in libcocos2d iOS.a(WebSocket.o)
  "_lws_context_destroy", referenced from:
      cocos2d::network::WebSocket::onSubThreadEnded() in libcocos2d iOS.a(WebSocket.o)
  "_lws_callback_on_writable", referenced from:
      cocos2d::network::WebSocket::onSubThreadLoop() in libcocos2d iOS.a(WebSocket.o)
      cocos2d::network::WebSocket::onConnectionOpened() in libcocos2d iOS.a(WebSocket.o)
  "_lws_service", referenced from:
      cocos2d::network::WebSocket::onSubThreadLoop() in libcocos2d iOS.a(WebSocket.o)
  "_lws_create_context", referenced from:
      cocos2d::network::WebSocket::onSubThreadStarted() in libcocos2d iOS.a(WebSocket.o)
  "_lws_is_final_fragment", referenced from:
      cocos2d::network::WebSocket::onClientReceivedData(void*, long) in libcocos2d iOS.a(WebSocket.o)
  "_lws_context_user", referenced from:
      cocos2d::network::WebSocketCallbackWrapper::onSocketCallback(lws*, lws_callback_reasons, void*, void*, unsigned long) in libcocos2d iOS.a(WebSocket.o)
  "_lws_client_connect", referenced from:
      cocos2d::network::WebSocket::onSubThreadStarted() in libcocos2d iOS.a(WebSocket.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by dumganhar** [comment]
You built it as i386 arch, it will definitely cause the linking problem for `x86_64` arch.

**Comment by div5yesh** [comment]
@dumganhar sorry my bad, the project does gets compiled when libraries built for x86_64 arch.

**Comment by div5yesh** [comment]
I'm still facing issues when connecting to wss://echo.websocket.org. It works fine for ws://echo.websocket.org.

```
Cocos2d-JS v3.14
2017-04-19 12:15:06.294 testproject-mobile[40227:3140212] cocos2d: surface size: 1704x960
libpng warning: iCCP: known incorrect sRGB profile
[WebSocket::init] _host: echo.websocket.org, _port: 80, _path: /
WebSocket thread start, helper instance: 0x6180000ba880
NOTICE: Initial logging level 967

NOTICE: Libwebsockets version: 2.1.0 divyeshch@LP3023.ho.games.com-v3.4-163-gc64b06b

NOTICE: IPV6 compiled in and enabled

NOTICE: libev support not compiled in

NOTICE: libuv support not compiled in

NOTICE:  Threads: 1 each 256 fds

NOTICE:  mem: platform fd map:  2048 bytes

NOTICE:  Compiled with OpenSSL support

NOTICE:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT

NOTICE: Creating Vhost 'default' port -1, 1 protocols, IPv6 on

NOTICE:  mem: per-conn:          512 bytes + protocol rx buf

NOTICE: lws_protocol_init

cocos2d: QuadCommand: resizing index size from [-1] to [2560]
ERR: SSL_new failed: error:00000001:lib(0):func(0):reason(1)

ERR: *** 336306371 error:140BA0C3:SSL routines:func(186):reason(195)

2017-04-19 12:15:06.612 testproject-mobile[40227:3140212] cocos2d: surface size: 1704x960
Couldn't send message since websocket wasn't opened!
```

**Comment by minggo** [comment]
@dumganhar  SSL support is added in v3.15, right?

**Comment by dumganhar** [comment]
@minggo , yes, it's supported from v3.15.
@div5yesh , could you try with v3.15(http://www.cocos2d-x.org/filedown/cocos2d-x-3.15.zip)
There are test case in ExtensionTests/WebSocketTest which also connects to `echo.websocke.org` with `wss` protocol. And we could make sure the test case works.

**Comment by div5yesh** [comment]
@dumganhar yes, I will surely try that and update this issue.

**Comment by div5yesh** [comment]
@dumganhar SSL socket connections are working fine with wss://echo.websocket.org without any issues.

But when I try to connect with my server, I get following errors:

D/WebSocket.cpp (879): ssl:1
D/WebSocket.cpp (837): WARNING: CA Root file isn't set. SSL connection will not peer server certificate
D/WebSocket.cpp (167): NOTICE: Creating Vhost 'default' port -1, 1 protocols, IPv6 on

D/WebSocket.cpp (895): port:443
D/WebSocket.cpp (167): NOTICE: lws_protocol_init
D/WebSocket.cpp (1326): WebSocket (0x7fb9f3805390) Unhandled websocket event: 32
D/WebSocket.cpp (1326): WebSocket (0x7fb9f3805390) Unhandled websocket event: 29
cocos2d: QuadCommand: resizing index size from [-1] to [2560]
2017-05-07 15:08:28.725 serious15-mobile[26568:7019880] cocos2d: surface size: 1704x960
D/WebSocket.cpp (167): **ERR: server's cert didn't look good, X509_V_ERR = 20**: error:00000014:lib(0):func(0):reason(20)

D/WebSocket.cpp (1201): WebSocket (0x7fb9f3805390) onConnectionError, state: 0 ...
D/WebSocket.cpp (1230): WebSocket (0x7fb9f3805390) onConnectionClosed, state: 2 ...
D/WebSocket.cpp (1260): onConnectionClosed, WebSocket (0x7fb9f3805390) is closing by server.
D/WebSocket.cpp (1283): WebSocket (0x7fb9f3805390) onConnectionClosed DONE!
D/WebSocket.cpp (1326): WebSocket (0x7fb9f3805390) Unhandled websocket event: 45
D/WebSocket.cpp (1326): WebSocket (0x7fb9f3805390) Unhandled websocket event: 33
D/WebSocket.cpp (1230): WebSocket (0x7fb9f3805390) onConnectionClosed, state: 3 ...
error::{"type":"error"}
close::{"type":"close"}
D/WebSocket.cpp (545): In the destructor of WebSocket (0x7fb9f3805390)
D/WebSocket.cpp (565): before join ws thread
D/WebSocket.cpp (167): NOTICE: lws_context_destroy

D/WebSocket.cpp (425): WebSocket thread exit, helper instance: 0x6080000a2e80
D/WebSocket.cpp (567): after join ws thread

**Comment by keyurdoshi18** [comment]
@div5yesh I have same issue
have you got the solution of your SSL socket connections ?

**Comment by bugs1111** [comment]
@div5yesh I have same issue
have you got the solution of your SSL socket connections ?

**Comment by mplulu** [comment]
@div5yesh I have same issue
have you got the solution of your SSL socket connections ?

**Comment by div5yesh** [comment]
@keyurdoshi18 , @bugs1111 , @mplulu 
The issue has been resolved. I was able to build build libwebsocket with SSL enabled. Refer the solution here: https://github.com/cocos2d/cocos2d-x/issues/17343

Moreover, you will get certificate error which can be fixed by using valid server SSL certificates.

Finally, I went on with neovisionaries-websocket Java library and JSB calls from custom JS websocket wrapper, since javascript execution is paused by cocos when app is minimized.
